File: Medium/Min Edge Movements to Connect a Graph/min-edge-movements-to-connect-a-graph.py
from collections import deque

class Solution:
    def minEdgesReq(self, V, edges):
        
        # To connect V vertices, we need at least V - 1 edges.
        if len(edges) < V - 1:
            return -1

        # Build adjacency list.
        adj = [[] for _ in range(V)]
        vis = [False] * V

        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        extraedges = 0
        components = 0

        # Find one connected component using BFS.
        def bfs(node):
            vis[node] = True

            q = deque([node])

            nonlocal edges
            nonlocal vertices

            while q:

                u = q.popleft()

                # Each undirected edge is counted twice.
                edges += len(adj[u])

                # Count vertices in this component.
                vertices += 1

                for v in adj[u]:
                    if not vis[v]:
                        vis[v] = True
                        q.append(v)

        # Find all connected components.
        for i in range(V):

            if not vis[i]:

                # Reset counts for the current component.
                edges = 0
                vertices = 0

                components += 1

                bfs(i)

                # Every undirected edge was counted twice.
                actualedges = edges // 2

                # A component with 'vertices' nodes needs
                # vertices - 1 edges to remain connected.
                #
                # Any additional edges are redundant/extra edges.
                extraedges += actualedges - vertices + 1

        # To connect 'components' components,
        # we need components - 1 edges.
        requiredEdges = components - 1

        # If enough redundant edges are available,
        # we can connect all the components.
        return requiredEdges if extraedges >= requiredEdges else -1

File: Medium/Min Edge Movements to Connect a Graph/min-edge-movements-to-connect-a-graph.cpp
class Solution {
public:
    vector<vector<int>> adj;
    vector<bool> vis;

    // BFS to find the number of vertices and edges
    // present in one connected component.
    void bfs(int u, int &vertices, int &edges) {

        queue<int> q;
        q.push(u);
        vis[u] = true;

        while (!q.empty()) {

            int cur = q.front();
            q.pop();

            // Count vertices in this component.
            vertices++;

            // Each undirected edge is counted twice.
            edges += adj[cur].size();

            for (int v : adj[cur]) {

                if (!vis[v]) {
                    vis[v] = true;
                    q.push(v);
                }
            }
        }
    }

    int minEdgesReq(int V, vector<vector<int>>& edges) {

        // To connect V vertices, at least V-1 edges are required.
        if (edges.size() < V - 1)
            return -1;

        // Build adjacency list.
        adj.resize(V);
        vis.assign(V, false);

        int extraedges = 0;
        int components = 0;

        for (auto &vec : edges) {

            int u = vec[0];
            int v = vec[1];

            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        // Find all connected components.
        for (int i = 0; i < V; i++) {

            if (!vis[i]) {

                components++;

                int vertices = 0;
                int edgess = 0;

                bfs(i, vertices, edgess);

                // Since the graph is undirected, every edge
                // was counted twice during BFS.
                int actualedges = edgess / 2;

                // A connected component with 'vertices' nodes
                // needs exactly vertices - 1 edges to form a tree.
                //
                // Any edge beyond that is an extra/redundant edge
                // that can potentially be used to connect components.
                extraedges += actualedges - vertices + 1;
            }
        }

        // To connect 'components' components,
        // we need components - 1 edges.
        int required = components - 1;

        // If we have enough redundant edges, we can connect
        // all the components.
        if (extraedges >= required)
            return required;

        return -1;
    }
};